File: alu/vk/src/main/kotlin/Alu.kt
import verik.base.*
import verik.data.*

class Alu: Module() {

    @input  var op  = t_Op()
    @input  var a   = t_UbitData()
    @input  var b   = t_UbitData()
    @output var x   = t_UbitData()

    @com fun update() {
        x = when (op) {
            Op.ADD -> a + b
            Op.SUB -> a - b
            Op.AND -> a and b
            Op.OR -> a or b
            Op.XOR -> a xor b
            Op.SLT -> if (s(a) < s(b)) u(1) else u(0)
            Op.SLTU -> if (a < b) u(1) else u(0)
            Op.SLL -> a shl b
            Op.SRL -> a shr b
            Op.SRA -> u(s(a) shr b)
        }
    }
}

File: alu/ch/src/main/scala/alu/Alu.scala
package alu

import chisel3._
import chisel3.stage.ChiselStage
import chisel3.util._

class Alu extends Module {

    val io = IO(new AluBundle)

    io.x := 0.U
    switch(io.op) {
        is (Op.add) { io.x := io.a + io.b }
        is (Op.sub) { io.x := io.a - io.b }
        is (Op.and) { io.x := io.a & io.b }
        is (Op.or) { io.x := io.a | io.b }
        is (Op.xor) { io.x := io.a ^ io.b }
        is (Op.slt) { io.x := Mux(io.a.asSInt() < io.b.asSInt(), 1.U, 0.U) }
        is (Op.sltu) { io.x := Mux(io.a < io.b, 1.U, 0.U) }
        is (Op.sll) { io.x := io.a << io.b }
        is (Op.srl) { io.x := io.a >> io.b }
        is (Op.sra) { io.x := (io.a.asSInt() >> io.b).asUInt() }
    }
}

object Main extends App {
    (new ChiselStage).emitVerilog(new Alu, args)
}
